app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java

```diff
@@ -328,7 +325,36 @@ public boolean onQueryTextChange(String text) {
             }
         });
 
+        // Set various initial states
         updateMenuToggleStates(_document.getFormat());
+
+        _saveMenuItem.setEnabled(_isTextChanged).getIcon().mutate().setAlpha(_isTextChanged ? 255 : 40);
+
+        final boolean canUndo = _editTextUndoRedoHelper.getCanUndo();
+        _undoMenuItem.setEnabled(canUndo).getIcon().mutate().setAlpha(canUndo ? 255 : 40);
+
+        final boolean canRedo = _editTextUndoRedoHelper.getCanRedo();
+        _redoMenuItem.setEnabled(canRedo).getIcon().mutate().setAlpha(canRedo ? 255 : 40);
+    }
+
+    private boolean _lastCanUndo = true;
+    private boolean _lastCanRedo = true;
+    private void updateUndoRedoIconStates() {
+        if (_editTextUndoRedoHelper == null) {
+            return;
+        }
+
+        final boolean canUndo = _editTextUndoRedoHelper.getCanUndo();
+        if (_undoMenuItem != null && _lastCanUndo != canUndo) {
+            _undoMenuItem.setEnabled(canUndo).getIcon().mutate().setAlpha(canUndo ? 255 : 40);
+            _lastCanUndo = canUndo;
+        }
+
+        final boolean canRedo = _editTextUndoRedoHelper.getCanRedo();
+        if (_redoMenuItem != null && _lastCanRedo != canRedo) {
+            _redoMenuItem.setEnabled(canRedo).getIcon().mutate().setAlpha(canRedo ? 255 : 40);
+            _lastCanRedo = canUndo;
+        }
```

Comment:
Breaking the undo / redo menu updates out so that we don't call redraw menu _every time_

Reply:
see the GsFragmentBase comment